packages/remirror__extension-entity-reference/src/entity-reference-extension.ts

```diff
@@ -0,0 +1,247 @@
+  CommandFunction,
+  CreateExtensionPlugin,
+  EditorState,
+  extension,
+  getTextSelection,
+  Helper,
+  helper,
+  isElementDomNode,
+  Mark,
+  MarkExtension,
+  MarkExtensionSpec,
+  MarkSpecOverride,
+  PrimitiveSelection,
+  ProsemirrorNode,
+  Transaction,
+  uniqueId,
+  within,
+} from '@remirror/core';
+import { Node } from '@remirror/pm/model';
+import { DecorationSet } from '@remirror/pm/view';
+
+import { EntityReferenceMetaData, EntityReferenceOptions } from './types';
+import { decorateEntityReferences } from './utils/decorate-entity-references';
+import { getDisjoinedEntityReferencesFromNode } from './utils/disjoined-entity-references';
+import { joinDisjoinedEntityReferences } from './utils/joined-entity-references';
+
+/**
+ *  Required props to create entityReference marks decorations.
+ */
+interface StateProps {
+  extension: EntityReferenceExtension;
+  state: EditorState;
+}
+
+interface EntityReferenceState {
+  entityReferences: { [key: string]: any };
+}
+
+const getEntityReferencesFromPluginState = (props: StateProps): EntityReferenceMetaData[][] => {
+  const { extension, state } = props;
+  const { entityReferences } = extension.getPluginState(state);
+  return entityReferences;
+};
+
+/**
+ * Creates a decoration set from the passed through state.
+ *
+ * @param props - see [[`DecorationProps`]]
+ */
+const createDecorationSet = (props: StateProps) => {
+  const { extension, state } = props;
+  const entityReferences = getEntityReferencesFromPluginState(props);
+  return DecorationSet.create(state.doc, extension.options.getStyle(entityReferences));
+};
+
+@extension<EntityReferenceOptions>({
+  defaultOptions: {
+    getStyle: decorateEntityReferences,
+    blockSeparator: undefined,
+  },
+})
+export class EntityReferenceExtension extends MarkExtension<EntityReferenceOptions> {
+  get name(): string {
+    return 'entity-reference' as const;
+  }
+
+  createMarkSpec(extra: ApplySchemaAttributes, override: MarkSpecOverride): MarkExtensionSpec {
+    const idAttr = `data-entity-reference`;
+    return {
+      ...override,
+      excludes: '',
+      attrs: {
+        ...extra.defaults(),
+        id: { default: '' },
+      },
+      toDOM: (mark: Mark) => [
+        'span',
+        {
+          ...extra.dom(mark),
+          [idAttr]: mark.attrs.id,
+        },
+        0,
+      ],
+      parseDOM: [
+        {
+          tag: `span[${idAttr}]`,
+          getAttrs: (node) => {
+            if (!isElementDomNode(node)) {
+              return false;
+            }
+
+            const id = node.getAttribute(idAttr);
+            return { ...extra.parse(node), id };
+          },
+        },
+        ...(override.parseDOM ?? []),
+      ],
+    };
+  }
+
+  /**
+   * Create the extension plugin for inserting decorations into the editor.
+   */
+  createPlugin(): CreateExtensionPlugin<EntityReferenceState> {
+    return {
+      state: {
+        init: (_: { [key: string]: any }, state: EditorState) => {
```

Comment:
```suggestion
        init: (_: EntityReferenceState, state: EditorState) => {
```

Reply:
Hmm, Typescript doesn't like this:
![image](https://user-images.githubusercontent.com/9339055/173194940-4f6c1866-fce9-479c-be9e-febf87c10e09.png)

Do you know why?

Reply:
@ronnyroeller I just wanted to have a quick look into this before we merge. Happy if I push to this branch?

Reply:
Sure, go ahead